### PR TITLE
Add pyright

### PR DESCRIPTION
**Pull request by drknzz**
### **Please check if the PR fulfills these requirements**
- [ ] The formatter, linter, and tests all run without an error
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

## **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -90,7 +92,7 @@ build-backend = "setuptools.build_meta"
 
 
 [tool.black]
-target-version = ["py37"]
+target-version = ["py38"]
```
Comment (cptartur) [inline pyproject.toml:95]: Good catch 👍

File: pyproject.toml [hunk 2]
```diff
@@ -100,3 +102,11 @@ markers = [
     "run_on_testnet: marks test that will only run on testnet (when --net=testnet)",
     "run_on_devnet: marks test that will only run on devnet (when --net=devnet)"
 ]
+
+[tool.pyright]
+include = ["starknet_py/"]
+exclude = [
+    "**/__pycache__",
+    "starknet_py/tests",
+    "**/*_test.py",
```
Comment (cptartur) [inline pyproject.toml:111]: Why aren't we typechecking tests?

File: starknet_py/net/client_models.py [hunk 3]
```diff
@@ -178,7 +178,7 @@ class DeclareTransactionResponse(SentTransactionResponse):
     Dataclass representing a result of declaring a contract on starknet
     """
 
-    class_hash: int = None
```
Comment (cptartur) [inline starknet_py/net/client_models.py:181]: I wonder if we should be having default values here? Perhaps this should be resolved on schema level instead?

File: starknet_py/net/client_models.py [hunk 4]
```diff
@@ -187,7 +187,7 @@ class DeployTransactionResponse(SentTransactionResponse):
     Dataclass representing a result of deploying a contract to starknet
     """
 
-    contract_address: int = None
```
Comment (cptartur) [inline starknet_py/net/client_models.py:190]: Same here

File: starknet_py/net/gateway_client.py [hunk 5]
```diff
@@ -96,7 +98,7 @@ async def get_block(
         res = await self._feeder_gateway_client.call(
             method_name="get_block", params=block_identifier
         )
-        return StarknetBlockSchema().load(res, unknown=EXCLUDE)
+        return StarknetBlockSchema().load(res, unknown=EXCLUDE)  # pyright: ignore
```
Comment (cptartur) [inline starknet_py/net/gateway_client.py:101]: Why are these ignores needed? Perhaps something like 
```py
wrapped_res: StarknetBlock = StarknetBlockSchema().load(res, unknown=EXCLUDE)
return wrapped_res
```
or a usage of `typing.cast` would be a better solution?

File: starknet_py/net/http_client.py [hunk 6]
```diff
@@ -22,7 +22,7 @@ def __init__(self, url, session: Optional[aiohttp.ClientSession] = None):
     def http_session(self) -> ClientSession:
         if self.session is not None:
             # noinspection PyTypeChecker
-            return nullcontext(self.session)
+            return nullcontext(self.session)  # pyright: ignore
```
Comment (cptartur) [inline starknet_py/net/http_client.py:25]: We should add a issue about covering this line with tests

File: starknet_py/net/http_client.py [hunk 7]
```diff
@@ -37,8 +37,8 @@ async def request(
                 session=session,
                 address=address,
                 http_method=http_method,
-                params=params,
-                payload=payload,
+                params=params,  # pyright: ignore
+                payload=payload,  # pyright: ignore
```
Comment (cptartur) [inline starknet_py/net/http_client.py:41]: Why are these needed here?

File: starknet_py/utils/data_transformer/data_transformer.py [hunk 8]
```diff
@@ -255,16 +278,20 @@ def _to_python_named(self, cairo_type, name, values):
...
             cairo_type.typ,
             cairo_type.name or "",
             value,
         )
         return value
 
-    def to_python(self, cairo_type: UsedCairoType, name: str, values: CairoData):
+    def to_python(
+        self, cairo_type: TypeTuple.Item, name: str, values: CairoData
+    ) -> PythonResult[Any]:
```
Comment (cptartur) [inline starknet_py/utils/data_transformer/data_transformer.py:294]: Couldn't some more specific typing be used here instead of Any?

File: starknet_py/net/l1/contracts.py [hunk 9]
```diff
@@ -36,7 +37,7 @@ def __init__(
         )
 
     def l2_to_l1_messages(
-        self, msg_hash: bytes, block_number: EthBlockIdentifier = None
+        self, msg_hash: bytes, block_number: typing.Optional[EthBlockIdentifier] = None
```
Comment (cptartur) [inline starknet_py/net/l1/contracts.py:40]: Why not import `Optional` and use it without `typing.`?

**Review by war-in — APPROVED** [review]

**Review by cptartur — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 2 above)
Comment (Solpatium) [reply pyproject.toml:111]: There is no point in doing this currently.

File: starknet_py/net/client_models.py (hunk 3 above)
Comment (war-in) [reply starknet_py/net/client_models.py:181]: SentTransactionResponse has default value so DeclareTransactionResponse must have it too

File: pyproject.toml (hunk 2 above)
Comment (cptartur) [reply pyproject.toml:111]: This should be mentioned in separate issue or in #390 then.

File: starknet_py/net/gateway_client.py (hunk 5 above)
Comment (Solpatium) [reply starknet_py/net/gateway_client.py:101]: It's more code, I am ok with having ignores here.

File: starknet_py/net/http_client.py (hunk 7 above)
Comment (Solpatium) [reply starknet_py/net/http_client.py:41]: Becuse they don't use None as accepted type even though they default to it below.

File: starknet_py/net/client_models.py (hunk 3 above)
Comment (cptartur) [reply starknet_py/net/client_models.py:181]: Could't it be changed? I'm not 100% sure what's the point of having keeping defaults in dataclasses instead of schemas.

File: starknet_py/net/l1/contracts.py (hunk 9 above)
Comment (Solpatium) [reply starknet_py/net/l1/contracts.py:40]: This module used typing as module before, I didn't want to change more lines than necessary.

File: starknet_py/net/gateway_client.py (hunk 5 above)
Comment (cptartur) [reply starknet_py/net/gateway_client.py:101]: This doesn't seem ideal, but i guess we could keep it. However `typing.cast` doesn't add that much code imo.

File: starknet_py/net/l1/contracts.py (hunk 9 above)
Comment (cptartur) [reply starknet_py/net/l1/contracts.py:40]: I'd change it anyways, it'd be more in line with what are we doing in pretty much every other file in the package.

File: starknet_py/net/client_models.py (hunk 3 above)
Comment (Solpatium) [reply starknet_py/net/client_models.py:181]: image

File: pyproject.toml (hunk 1 above)
Comment (Solpatium) [reply pyproject.toml:95]: @drknzz 👍

File: starknet_py/net/client_models.py (hunk 3 above)
Comment (Solpatium) [reply starknet_py/net/client_models.py:181]: You can change this in a seperate PR if you want. This PR has a different purpose.

File: pyproject.toml (hunk 2 above)
Comment (Solpatium) [reply pyproject.toml:111]: I am not sure if we need to typecheck tests at all tbh

File: starknet_py/net/http_client.py (hunk 6 above)
Comment (Solpatium) [reply starknet_py/net/http_client.py:25]: https://github.com/software-mansion/starknet.py/issues/391

File: starknet_py/net/client_models.py (hunk 4 above)
Comment (Solpatium) [reply starknet_py/net/client_models.py:190]: Let's discuss this in thread above.

File: pyproject.toml (hunk 2 above)
Comment (Solpatium) [reply pyproject.toml:111]: I created an issue, but added a low priority https://github.com/software-mansion/starknet.py/issues/392

File: starknet_py/net/client_models.py (hunk 3 above)
Comment (cptartur) [reply starknet_py/net/client_models.py:181]: Okay, this can be resolved in a separate PR

**Review by cptartur — APPROVED** [review]
